### PR TITLE
Signed-Peer-Record support in Pubsub/Gossipsub message transfer

### PR DESCRIPTION
**Pull request by lla-dane**
Tracks #791 

This is a WIP PR for adding signed-peer-record transfer in pubsub message transfer for ore secure and trustless communication between peers.

Changes:
Pubsub RPC Message Schema updated:
- `optional bytes senderRecord` -  contains the sender's signed peer-record

```
message RPC {
	repeated SubOpts subscriptions = 1;
	repeated Message publish = 2;

	message SubOpts {
		optional bool subscribe = 1; // subscribe or unsubscribe
		optional string topicid = 2;
	}

	optional ControlMessage control = 3;
	optional bytes senderRecord = 4; // Envelope(PeerRecord) encoded
}
```
Everytime a new message is received or sent out:

- When a message is sent out: sender's(local-peer) fresh signed-peer-record is attached to the message protobuf, the signed-peer-record in the response message is consumed and populated in the Peerstore

- When a message is received: The sender's signed-peer-record is consumed and populated in the PeerStore, and the in the response message a fresh signed-peer-record is attached to response message protobuf.

RFC related to signed-envelope and peer-routing-records:
[libp2p/specs@master/RFC/0002-signed-envelopes.md](https://github.com/libp2p/specs/blob/master/RFC/0002-signed-envelopes.md?rgh-link-date=2025-08-09T12%3A17%3A41Z)
[libp2p/specs@master/RFC/0003-routing-records.md](https://github.com/libp2p/specs/blob/master/RFC/0003-routing-records.md?rgh-link-date=2025-08-09T12%3A17%3A41Z)

### CODE REVIEW

**Comment by lla-dane** [comment]
The basic implementation for sending and receiving signed-peer-records and processing them via appropriate utilities in the Certified-Addr-Book interface has been added.

What remains is:

1. Writing tests, to verify that the records are sent/received properly
2. Maybe prioritizing the acceptance of a PubSub messages on the basis of valid signed-peer-records, meaning in cases like, the senderRecord (which is signed-peer-record of the sender) that is received in the Pubsub RPC message is invalid (signature in the envelope is invalid corresponding to the peer-record that we received => Meaning the envelope has been tampered with after signing), so in this case, do you process the pubsub-rpc message or reject it ?

@pacrob @seetadev, please take a look and provide feedback.

**Comment by pacrob** [comment]
See my comments in #815 regarding excessive SPR creation and peer ID validation.

> in this case, do you process the pubsub-rpc message or reject it ?

Is there any reason we would not want to reject a record that fails checks?

**Comment by lla-dane** [comment]
Now remains: Adding tests and rejecting the rpc messages sending invalid signed-records

**Comment by seetadev** [comment]
@lla-dane : Neat work. Thank you for addressing the feedback points shared by @pacrob. Appreciate it.

Looking forward to seeing this key feature enhancement. Will ask @Winter-Soren to test it out in Gossipsub 1.2 branch.

File: tests/core/pubsub/test_pubsub.py [hunk 1]
```diff
@@ -95,11 +137,52 @@ async def test_peers_subscribe():
...
+        if envelope is not None:
+            envelope.public_key = key_pair.public_key
+            pubsubs_fsub[0].host.get_peerstore().set_local_record(envelope)
+
+        await pubsubs_fsub[0].subscribe(TESTING_TOPIC)
+        # Yeild to let 0 notify 1
+        await trio.sleep(1)
+        assert pubsubs_fsub[0].my_id not in pubsubs_fsub[1].peer_topics.get(
+            TESTING_TOPIC, set()
+        )
```
Comment (lla-dane) [inline tests/core/pubsub/test_pubsub.py]: Added the mechanism for ignoring pubsub messages upon receiving invalid-signed-records, and here's the test for it. @pacrob @seetadev

**Comment by lla-dane** [comment]
This PR is ready for further review. 
Caching the latest signed-record for local node (for re-issuing), and rejecting the pubsub messages upon receiving invalid-records has been added.

File: tests/core/pubsub/test_pubsub.py (hunk 1 above)
Comment (seetadev) [reply tests/core/pubsub/test_pubsub.py]: @lla-dane : Great addition, Abhinav — this test really strengthens the safety guarantees around pubsub. 

It’s exactly the kind of edge case we need to cover: making sure that when a peer’s signed record is corrupted or invalid, subscriptions and message propagation don’t proceed silently.

Thanks for pushing this through with a concrete test case. It’ll make the pubsub layer more robust and give us confidence that invalid records are caught and handled as expected. Looking forward to seeing this merged in with the rest of the SPR verification improvements. Excellent work!

**Comment by seetadev** [comment]
> This PR is ready for further review.
> Caching the latest signed-record for local node (for re-issuing), and rejecting the pubsub messages upon receiving invalid-records has been added.

Great work, @lla-dane — thanks for pushing this PR across the finish line. Will do a final review and share feedback soon. Will also look forward to receiving feedback from @pacrob. 

This PR is heading towards final review + merge.

**Comment by pacrob** [comment]
This PR has several of the same concerns that have been noted in #815. Please update there and apply the same updates here. I will then be able to review further.

**Comment by lla-dane** [comment]
@pacrob: I have done most of the changes that were suggested in #815 in 3d0b6a7
- Moved `env_to_send_in_RPC` to `libp2p/peer/peerstore.py`.
- Commented the `bool` variable from `env_to_send_in_RPC` where-ever it was used.
- Now using `env_to_send_in_RPC` for record issuing in identify.
- We check the sender's peer-id in `maybe_consume_peer_record` function against the sent peer-record in `libp2p/pubsub/utils.py` file, and test is appeneded in `test_pubsub.py::test_peer_subscribe_fail_upon_invald_record_transfer`

Please suggest any more improvements that could be don.

File: tests/core/pubsub/test_pubsub.py [hunk 2]
```diff
@@ -87,6 +91,45 @@ async def test_re_unsubscribe():
...
+            .host.get_peerstore()
+            .get_peer_record(pubsubs_fsub[0].host.get_id())
+        )
+        assert isinstance(envelope_b_unsub, Envelope)
+
+        # This proves that a freshly signed record was issued rather than
+        # the latest-cached-one creating one.
+        assert envelope_b_sub.record().seq < envelope_b_unsub.record().seq
+
+
```
Comment (lla-dane) [inline tests/core/pubsub/test_pubsub.py:132]: Also @pacrob, about the tests for this PR -- the flow of signed records is already covered in the `subscribe/unsubscribe` executions in floodsub. Do I also need to simulate the transfers in the other executions, given that the pattern is the same across all of them?

**Comment by seetadev** [comment]
@lla-dane : Thank you for making improvements as suggested by @pacrob. Appreciate your efforts.

**Comment by lla-dane** [comment]
Added newsfragment

**Comment by seetadev** [comment]
@lla-dane : Thank you so much for making all the requisite improvements. 

Once I will get a thumbs up from @pacrob, will merge the PR. 

Doing a final review on the code commits added. Great work indeed.

**Comment by seetadev** [comment]
@lla-dane :  It would be great if you could share a detailed write-up (maybe as a [discussion post](https://github.com/libp2p/py-libp2p/discussions)) on how you approached this PR, the steps you followed, and any challenges or insights you had.

Reflections like these will help new pub-sub and gossipsub contributors learn from your practical experiences in this important PR and give us a stronger record of how things evolve in the py-libp2p project.

**Comment by pacrob** [comment]
@seetadev Let's get #815 merged first. This PR contains some of the same setup code from that PR, so better to be able to see just the new changes.

File: tests/core/pubsub/test_pubsub.py (hunk 2 above)
Comment (pacrob) [reply tests/core/pubsub/test_pubsub.py:132]: If you're testing the same code (as in, imported from the same place), then one test is fine. If it's similar code that is located in multiple places, each location should be tested.

**Comment by lla-dane** [comment]
@seetadev, @pacrob: Fixed the merge conflicts here. Please see if anything else needs to be done in this PR.

**Comment by seetadev** [comment]
@lla-dane : Great work, Abhinav. Appreciate your efforts.

Also, it would be great if you could share a detailed write-up (maybe as a [discussion post](https://github.com/libp2p/py-libp2p/discussions)) on how you approached this PR, the steps you followed, and any challenges or insights you had.

Reflections like these will help new pub-sub and gossipsub contributors learn from your practical experiences in this important PR and give us a stronger record of how things evolve in the py-libp2p project.

Once I will get a thumbs up from @pacrob, will merge the PR. 

Doing a final review on the code commits added. Wonderful contribution indeed.

**Comment by lla-dane** [comment]
This PR is continued at #889
